src/main/java/org/mule/service/http/impl/service/server/grizzly/ResponseStreamingCompletionHandler.java

```diff
@@ -45,15 +57,39 @@ public ResponseStreamingCompletionHandler(final FilterChainContext ctx, final Ht
     httpResponsePacket = buildHttpResponsePacket(request, httpResponse);
     inputStream = httpResponse.getEntity().getContent();
     memoryManager = ctx.getConnection().getTransport().getMemoryManager();
+    bufferSize = calculateBufferSize(ctx, httpResponse);
+    LOGGER.debug("Response streaming calculated buffer size = {} bytes.");
     this.responseStatusCallback = responseStatusCallback;
   }
 
+  private int calculateBufferSize(FilterChainContext ctx, HttpResponse httpResponse) {
+    String contentLengthHeader = httpResponse.getHeaderValue(CONTENT_LENGTH);
+    int contentLength;
+    if (!isEmpty(contentLengthHeader)) {
+      contentLength = Integer.valueOf(contentLengthHeader);
+    } else {
+      contentLength = -1;
+    }
+    if (contentLength > 0) {
+      try {
+        LOGGER.debug("Content length header present, calculating maximal buffer size.");
+        return min(MAX_SEND_BUFFER_SIZE, min(ctx.getConnection().getWriteBufferSize(), inputStream.available()));
```

Comment:
Using the stream seems odd since you'll calculate it once and use it in all chunks. If you're trying to match the PipedInputStream buffer, are you sure this will work?

Reply:
i agree with Ana here, i don't think there's any guaranty that the read bytes will be consistent across chunks.